crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1897 @@
+pub struct MerkleBlob {
+    blob: Vec<u8>,
+    free_indexes: HashSet<TreeIndex>,
+    key_to_index: HashMap<KvId, TreeIndex>,
+}
+
+impl MerkleBlob {
+    pub fn new(blob: Vec<u8>) -> Result<Self, Error> {
+        let length = blob.len();
+        let remainder = length % BLOCK_SIZE;
+        if remainder != 0 {
+            return Err(Error::InvalidBlobLength(remainder));
+        }
+
+        let (free_indexes, key_to_index) = get_free_indexes_and_keys_values_indexes(&blob);
+
+        Ok(Self {
+            blob,
+            free_indexes,
+            key_to_index,
+        })
+    }
+
+    fn clear(&mut self) {
+        self.blob.clear();
+        self.key_to_index.clear();
+        self.free_indexes.clear();
+    }
+
+    pub fn insert(
+        &mut self,
+        key: KvId,
+        value: KvId,
+        hash: &Hash,
+        insert_location: InsertLocation,
+    ) -> Result<(), Error> {
+        if self.key_to_index.contains_key(&key) {
+            return Err(Error::KeyAlreadyPresent);
+        }
+
+        let insert_location = match insert_location {
+            InsertLocation::Auto {} => self.get_random_insert_location_by_kvid(key)?,
+            _ => insert_location,
+        };
+
+        match insert_location {
+            InsertLocation::Auto {} => {
+                unreachable!("this should have been caught and processed above")
+            }
+            InsertLocation::AsRoot {} => {
+                if !self.key_to_index.is_empty() {
+                    return Err(Error::UnableToInsertAsRootOfNonEmptyTree);
+                };
+                self.insert_first(key, value, hash)?;
+            }
+            InsertLocation::Leaf { index, side } => {
+                let old_leaf = self.get_node(index)?.try_into_leaf()?;
+
+                let internal_node_hash = match side {
+                    Side::Left => internal_hash(hash, &old_leaf.hash),
+                    Side::Right => internal_hash(&old_leaf.hash, hash),
+                };
+
+                let node = LeafNode {
+                    parent: None,
+                    hash: *hash,
+                    key,
+                    value,
+                };
+
+                if self.key_to_index.len() == 1 {
+                    self.insert_second(node, &old_leaf, &internal_node_hash, side)?;
+                } else {
+                    self.insert_third_or_later(node, &old_leaf, index, &internal_node_hash, side)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn insert_first(&mut self, key: KvId, value: KvId, hash: &Hash) -> Result<(), Error> {
+        let new_leaf_block = Block {
+            metadata: NodeMetadata {
+                node_type: NodeType::Leaf,
+                dirty: false,
+            },
+            node: Node::Leaf(LeafNode {
+                parent: None,
+                key,
+                value,
+                hash: *hash,
+            }),
+        };
+
+        self.clear();
```

Comment:
it's not obvious to me why `insert_first()` clears everything in `Self`, along with the other functions that also `clear()` self.

Reply:
the call here in `.insert_first()` is theoretically pointless.  it does stand a chance of cleaning up any state left in the caches, but it is likely quite uncommon that folks end up emptying their store anyways so this is probably more confusing than valuable at all.  removed in 7867000996e0824b2de457528e5cac091e54ab0d.

in `.add_second()` the existing code just starts a fresh tree and adds back the data from the single existing block.  when adding a second leaf, the first leaf is presently 'moved', a new internal node created as the root, and an additional leaf added.  basically, nothing is staying the same.  so, we could remove the existing node from the key to index cache etc, or we could just wipe 'everything' (one thing) and put stuff where we want it.

in `.delete()`, the use of `.clear()` occurs only when the last leaf is deleted and the tree is empty.  again, we could carefully clean each cache, or just wipe them all because we know there should be nothing.

i do kind of feel like there's opportunity for simplification of the various logic around inserting and deleting.  that may yield better patterns of code without these corners that use `.clear()` for simplicity.

Reply:
I see, so what's being cleared is a kind of cache of parsed tree structure. and when we change the view to look at a different part of the tree, we clear the cache and parse that part. is that right?